plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -626,7 +705,31 @@ enterComputeCek costs = computeCek where
       case resultOrErr of
           -- Turn the cause of a possible failure, being a 'CekValue', into a 'Term'.
           Left e       -> throwCek $ mapCauseInMachineException (void . dischargeCekValue) e
-          Right result -> returnCek ctx result
+          Right result -> returnCek unbudgetedSteps ctx result
+
+    -- | Spend the budget that has been accumulated for a number of machine steps.
+    spendAccumulatedBudget :: WordArray -> CekM s ()
+    spendAccumulatedBudget !unbudgetedSteps = iforWordArray unbudgetedSteps spend
+
+    -- Making this a definition of its own causes it to inline better than actually writing it inline, for
+    -- some reason.
+    -- Skip index 7, that's the total counter!
+    -- See Note [Structure of the step counter]
+    {-# INLINE spend #-}
+    spend !i !w = unless (i == 7) $ do
+        let kind :: StepKind = toEnum i
+        spendBudgetCek (BStep kind) (stimes w (cekStepCost ?cekCosts kind))
+
+    -- | Accumulate a step, and maybe spend the budget that has accumulated for a number of machine steps, but only if we've exceeded our slippage.
+    stepAndMaybeSpend :: StepKind -> WordArray -> CekM s WordArray
+    stepAndMaybeSpend kind unbudgetedSteps = do
+        -- See Note [Structure of the step counter]
+        let ix = fromIntegral $ fromEnum kind
+            unbudgetedSteps' = overIndex 7 (+1) $ overIndex ix (+1) unbudgetedSteps
+            unbudgetedStepsTotal = readArray unbudgetedSteps' 7
+        if unbudgetedStepsTotal >= ?cekSlippage
+        then spendAccumulatedBudget unbudgetedSteps' >> pure (toWordArray 0)
+        else pure unbudgetedSteps'
```

Comment:
Perhaps worth an explicit comment that `unbudgetedSteps'` is forced, because `unbudgetedStepsTotal` is forced by the comparison and is obtained from `unbudgetedSteps'`.

Reply:
I'm just going to put more bangs. It can't hurt.